daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/ImmArray.scala

```diff
@@ -354,33 +374,80 @@ final class ImmArray[+A] private (
 }
 
 object ImmArray extends ImmArrayInstances {
+
+  class Builder[T](initCapacity: Int = 0) extends mutable.Builder[T, ImmArray[T]] {
+    var size: Int = 0
+    var buff: Array[AnyRef] = new Array(initCapacity)
+
+    private[this] def increaseCapacity(addedCapacity: Int): Unit = {
+      val newCapacity = buff.length + addedCapacity
+      val newBuff = new Array[AnyRef](newCapacity)
+      System.arraycopy(buff, 0, newBuff, 0, size)
+      buff = newBuff
+    }
+
+    override def sizeHint(i: Int): Unit = {
+      val addedCapacity = size + i - buff.length
+      if (addedCapacity > 0)
+        increaseCapacity(addedCapacity)
+    }
+
+    override def +=(elem: T): this.type = {
+      if (size == buff.length) increaseCapacity(size max 1)
+      buff(size) = elem.asInstanceOf[AnyRef]
+      size += 1
+      this
+    }
+
+    override def clear(): Unit =
+      size = 0
+
+    override def result(): ImmArray[T] =
+      if (size == 0)
+        ImmArray.empty
+      else {
+        val immArray = new ImmArray(0, size, buff)
+        // to be sure the builder and `buff` are not reused.
+        buff = null
+        immArray
+      }
+  }
+
   private[this] val emptySingleton: ImmArray[Nothing] =
-    ImmArray.unsafeFromArraySeq(new mutable.ArraySeq(0))
+    new ImmArray(0, 0, Array.empty)
   def empty[T]: ImmArray[T] = emptySingleton
 
   def apply[T](element0: T, elements: T*): ImmArray[T] = {
-    val builder = ImmArray.newBuilder[T]
+    val builder = new Builder[T](1)
     builder += element0
+    builder.sizeHint(elements)
```

Comment:
Shouldn't this be
```suggestion
    builder.increaseCapacity(elements)
```
or better
```
val builder = new Builder[T](1 + elements)
```
above?

Reply:
fixed.